flink-runtime/src/main/java/org/apache/flink/runtime/rest/FileUploadHandler.java

```diff
@@ -89,7 +89,7 @@ protected void channelRead0(final ChannelHandlerContext ctx, final HttpObject ms
 			final HttpContent httpContent = (HttpContent) msg;
 			currentHttpPostRequestDecoder.offer(httpContent);
 
-			while (currentHttpPostRequestDecoder.hasNext()) {
```

Comment:
The failure was triggered every-time i attempted to upload anything, but i may very well depend on the payload size.

Your analysis is mostly correct, what's missing is that decoder switches into the `EPILOGUE` state when being offered a `LastHttpContent`.
If this last message is not empty the exception is not thrown since data is still available, checked in `hasNext` via `this.bodyListHttpDataRank >= this.bodyListHttpData.size()`.

Reply:
The `JobSubmitHandlerTest` never transmits any files via netty. The failure would occur in the `FileUploadHandler` which is currently completely untested.

I'll try to find a test that can reproduce this